### PR TITLE
Algorithm nice names

### PR DESCRIPTION
**Pull request by williamcroberts**
This adds support for both PCR lists and algorithm specifiers to use human friendly names rather than hex. I also organized the man pages a little but so the redundant what algorithms are supported are grouped into common files. System tests were updated to use the new nice names.

TODO: Add unit tests.

### CODE REVIEW

File: test/system/test_tpm2_createpolicy.sh [hunk 1]
```diff
@@ -35,23 +35,21 @@
 
 declare -A expected_policy_digest=(["sha1"]="f28230c080bbe417141199e36d18978228d8948fc10a6a24921b9eba6bb1d988"
                                    ["sha256"]="33e36e786c878632494217c3f490e74ca0a3a122a8a4f3c5302500df3b32b3b8")
-declare -A hash_name=([0x4]="sha1" [0xb]="sha256")
 pcr_index=0
 
-for halg in 0x4 0xb
+for halg in sha1 sha256
 do
```
Comment (martinezjavier) [inline test/system/test_tpm2_createpolicy.sh]: In some tests I see that you use a mix of nice name and hex values and in other tests you convert all of them to nice names. I prefer the former to make sure that the tests exercise both options.

But I think there should be some consistency, either the tests should exercise both or all should be converted to nice names.

File: tools/tpm2_sign.c [hunk 2]
```diff
@@ -289,8 +290,8 @@ static bool init(int argc, char *argv[], tpm_sign_ctx *ctx) {
         }
     }
 
-    if (!((flags.k || flags.c) && flags.g && flags.m && flags.s)) {
-        LOG_ERR("Expected options (k or c) and g and m and s");
+    if (!((flags.k || flags.c) && flags.m && flags.s)) {
+        LOG_ERR("Expected options (k or c) and m and s");
         return false;
```
Comment (martinezjavier) [inline tools/tpm2_sign.c]: This is an unrelated fix (-g option isn't needed because sha1 is the default), so I think it should be in a separate patch with its own explanation why is needed.

File: tools/tpm2_verifysignature.c [hunk 3]
```diff
@@ -321,7 +322,7 @@ int execute_tool(int argc, char *argv[], char *envp[], common_opts_t *opts,
 
     tpm2_verifysig_ctx ctx = {
             .flags = { .all = 0 },
-            .halg = TPM_ALG_SHA256,
+            .halg = TPM_ALG_SHA1,
             .msgHash = TPM2B_TYPE_INIT(TPM2B_DIGEST, buffer),
```
Comment (martinezjavier) [inline tools/tpm2_verifysignature.c]: I think changing the default should also be split in a separate commit.

**Comment by martinezjavier** [comment]
@williamcroberts this is great! I've tested your branch and it works good for me. It also allowed me to remove code in a script that only had to convert from pretty names for users to the hex values to pass to the tools. Thanks!

I just had a few minor comments in case you want to take it into account.

File: Makefile.am [hunk 4]
```diff
@@ -175,6 +177,10 @@ test_unit_test_tpm2_nv_util_CFLAGS   = $(AM_CFLAGS) $(CMOCKA_CFLAGS)
 test_unit_test_tpm2_nv_util_LDADD    = $(CMOCKA_LIBS) $(LIB_COMMON)
 test_unit_test_tpm2_nv_util_SOURCES  = test/unit/test_tpm2_nv_util.c
 
+test_unit_test_tpm2_alg_util_CFLAGS   = $(AM_CFLAGS) $(CMOCKA_CFLAGS)
+test_unit_test_tpm2_alg_util_LDADD    = $(CMOCKA_LIBS) $(LIB_COMMON)
+test_unit_test_tpm2_alg_util_SOURCES  = test/unit/test_tpm2_alg_util.c
+
```
Comment (williamcroberts) [inline Makefile.am]: Yeah this is an artifact of me accidentally running git clean without adding the test files, I actually had them in place. So I want to re-create them and squash them into this commit.

File: lib/tpm2_alg_util.c [hunk 5]
```diff
@@ -0,0 +1,143 @@
+#include <string.h>
+#include <stdbool.h>
+#include <stdlib.h>
+
```
Comment (williamcroberts) [inline lib/tpm2_alg_util.c]: done

File: lib/tpm2_alg_util.c [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+#include <sapi/tpm20.h>
+
+#include "tpm2_util.h"
+#include "tpm2_alg_util.h"
+
```
Comment (williamcroberts) [inline lib/tpm2_alg_util.c]: done

File: lib/tpm2_alg_util.c [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+#endif
+#ifndef ALG_SHA3_384_VALUE
+#define ALG_SHA3_384_VALUE 0x28
+#endif
+#ifndef ALG_SHA3_512_VALUE
+#define ALG_SHA3_512_VALUE 0x29
+#endif
+
+void tpm2_alg_util_for_each_alg(tpm2_alg_util_alg_iteraror iterator, void *userdata) {
+
```
Comment (williamcroberts) [inline lib/tpm2_alg_util.c]: Well it is iterating over the contents, what you choose to do with it is arbitrary. I was torn between just making the data extern vs encapsulating it like this. The encapsulation here insn't buying us much except for avoiding the writing of a for loop for what that's worth.

File: lib/tpm2_alg_util.c [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+        { .name = "cbc", .id = ALG_CBC_VALUE },
+        { .name = "cfb", .id = ALG_CFB_VALUE },
+        { .name = "ecb", .id = ALG_ECB_VALUE },
+    };
+
+    size_t i;
+    for (i=0; i < ARRAY_LEN(algs); i++) {
+        const alg_pair *alg = &algs[i];
+        bool result = iterator(alg->id, alg->name, userdata);
+        if (result) {
```
Comment (williamcroberts) [inline lib/tpm2_alg_util.c]: sure, that name is not very good.

File: test/system/test_tpm2_certify.sh [hunk 9]
```diff
@@ -38,8 +38,9 @@ file_certify_key_name=name.load.B1_B8
 file_output_attest=attest.out
 file_output_signature=certify_signature.out 
   
-
+# do not update alg_hash without updating alg_hash_name
 alg_hash=0x000B
+alg_hash_name=sha256
```
Comment (williamcroberts) [inline test/system/test_tpm2_certify.sh]: I wanted to mix and match the hex values and the nice names to get coverage on both paths.

File: man/object-alg-common.troff [hunk 10]
```diff
@@ -0,0 +1,8 @@
+.br
+Supported public object algorithms are:
+  '0x1' or 'rsa' for TPM_ALG_RSA (default)
+  '0x8' or 'keyedhash' for TPM_ALG_KEYEDHASH
+  '0x23' or 'ecc' for TPM_ALG_ECC
+  '0x25' or 'sha512' for TPM_ALG_SHA512
+.br
```
Comment (williamcroberts) [inline man/object-alg-common.troff]: Good catch

File: lib/tpm2_alg_util.c [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+typedef struct convert_data convert_data;
+struct convert_data {
+    const char *name;
+    TPM_ALG_ID id;
+};
+
+static bool convert_iterator(TPM_ALG_ID id, const char *name, void *userdata) {
+
+    convert_data *d = (convert_data *)userdata;
+
```
Comment (williamcroberts) [inline lib/tpm2_alg_util.c]: Well if someone needs to pass other userdata along it makes it difficult. The fact that it's a convert_data structure is just because of how what is being done while iterating over the data elements.

File: tools/tpm2_verifysignature.c (hunk 3 above)
Comment (williamcroberts) [inline tools/tpm2_verifysignature.c]: When I update the man page for it, the default is mentioned in the common file, so if I split it into a separate commit, there's a point in time where the man page doesn't match reality, which I don't think we would ever want to happen. Splitting it doesn't buy us anything, because if we revert it, we'd have to revert the other patch as well.

File: Makefile.am (hunk 4 above)
Comment (martinezjavier) [inline Makefile.am]: Ok.

File: lib/tpm2_alg_util.c (hunk 7 above)
Comment (martinezjavier) [inline lib/tpm2_alg_util.c]: Yes, it's iterating but I think that's an implementation detail. From the user POV is finding a match for a given algorithm name / ID. But I don't have a strong opinion on this so I'm OK on leaving it as is.

File: test/system/test_tpm2_certify.sh (hunk 9 above)
Comment (martinezjavier) [inline test/system/test_tpm2_certify.sh]: Ok.

File: tools/tpm2_verifysignature.c (hunk 3 above)
Comment (martinezjavier) [inline tools/tpm2_verifysignature.c]: Ah, good point. I didn't think about the man page.

File: lib/tpm2_alg_util.c (hunk 11 above)
Comment (martinezjavier) [inline lib/tpm2_alg_util.c]: Yes, my point was that I don't see what other user data could be provided. Most likely alg_pair will be enough and if someone ever needs something different, then it could be made more generic.

File: lib/tpm2_alg_util.c (hunk 8 above)
Comment (williamcroberts) [inline lib/tpm2_alg_util.c]: You'll see how this indirection of a callback and void * is used in a more generic sense when you see the tests. In the tests, my callback for each element is searching a list and modifying it. Thus its iterator callback and userdata are unique to it.

I would like to modify a tool or tools so that they could pass an option to dump this map. Perhaps taking rc_decode, and modifying it to be able to decode algorithms and dump the map.

File: lib/tpm2_alg_util.c [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+        }
+    }
+}
+
+typedef struct convert_data convert_data;
+struct convert_data {
+    const char *name;
+    TPM_ALG_ID id;
+};
+
```
Comment (williamcroberts) [inline lib/tpm2_alg_util.c]: Fair enough.

File: lib/tpm2_alg_util.c (hunk 7 above)
Comment (williamcroberts) [inline lib/tpm2_alg_util.c]: The tests will use a different one, and then eventually some tool to dump the map will use one, eahc with there own task to be done on each element, and all will have different userdata.

File: lib/tpm2_alg_util.c [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+struct alg_pair {
+    const char *name;
+    TPM_ALG_ID id;
+};
+
+/*
+ * The TSS has a bug where it was missing algs 0x27 trough 0x29.
+ * see: https://github.com/01org/TPM2.0-TSS/issues/476
+ * per https://trustedcomputinggroup.org/wp-content/uploads/TCG_Algorithm_Registry_Rev_1.24.pdf
+ */
```
Comment (williamcroberts) [inline lib/tpm2_alg_util.c]: FIXME with bug: https://github.com/01org/tpm2.0-tools/issues/375

File: lib/tpm2_alg_util.c (hunk 5 above)
Comment (williamcroberts) [inline lib/tpm2_alg_util.c]: done

File: lib/tpm2_alg_util.c (hunk 6 above)
Comment (williamcroberts) [inline lib/tpm2_alg_util.c]: done

**Comment by williamcroberts** [comment]
@martinezjavier I think I got all your comments either addressed or fixed. Thank you for the review. I added the tests back in, that provide (I think) ample coverage. I'll leave this up for another day and see if any new feedback is left.

File: test/unit/test_tpm2_alg_util.c [hunk 14]
```diff
@@ -0,0 +1,277 @@
...
+        assert_int_equal(value, found_id); \
+        assert_int_equal(value, from_hex_str); \
+        assert_int_equal(value, from_nice_str); \
+    }
+
+single_item_test(rsa, ALG_RSA_VALUE)
+/*
+ * sha sha1 is it's own test, as alg to string
+ * can return either, based on the map ordering.
+ *
```
Comment (williamcroberts) [inline test/unit/test_tpm2_alg_util.c]: extra space, delete it.

File: test/unit/test_tpm2_alg_util.c [hunk 15]
```diff
@@ -0,0 +1,277 @@
...
+    ALG_SYMCIPHER_VALUE,
+    ALG_CAMELLIA_VALUE,
+    ALG_SHA3_256_VALUE,
+    ALG_SHA3_384_VALUE,
+    ALG_SHA3_512_VALUE,
+    ALG_CTR_VALUE,
+    ALG_OFB_VALUE,
+    ALG_CBC_VALUE,
+    ALG_CFB_VALUE,
+    ALG_ECB_VALUE };
```
Comment (williamcroberts) [inline test/unit/test_tpm2_alg_util.c]: formatting }; on last line.

File: lib/tpm2_alg_util.c (hunk 8 above)
Comment (martinezjavier) [inline lib/tpm2_alg_util.c]: @williamcroberts right, I looked and the tests and now I agree with the code. 

I just didn't want the indirection level if it would not be used (I didn't know the tests where using it since weren't pushed in the first version). Thanks for the clarification.
